src/detect.c

```diff
@@ -813,22 +813,25 @@ static inline void DetectRulePacketRules(
         DetectRunPostMatch(tv, det_ctx, p, s);
 
         uint64_t txid = PACKET_ALERT_NOTX;
-        if ((alert_flags & PACKET_ALERT_FLAG_STREAM_MATCH) ||
-                (s->alproto != ALPROTO_UNKNOWN && pflow->proto == IPPROTO_UDP)) {
-            // if there is a stream match (TCP), or
-            // a UDP specific app-layer signature,
-            // try to use the good tx for the packet direction
-            if (pflow->alstate) {
-                uint8_t dir =
-                        (p->flowflags & FLOW_PKT_TOCLIENT) ? STREAM_TOCLIENT : STREAM_TOSERVER;
-                txid = AppLayerParserGetTransactionInspectId(pflow->alparser, dir);
+        if (pflow && pflow->alstate) {
+            uint8_t dir = (p->flowflags & FLOW_PKT_TOCLIENT) ? STREAM_TOCLIENT : STREAM_TOSERVER;
+            txid = AppLayerParserGetTransactionInspectId(pflow->alparser, dir);
+            if ((s->alproto != ALPROTO_UNKNOWN && pflow->proto == IPPROTO_UDP) ||
+                    (de_ctx->force_applayer &&
+                            AppLayerParserGetTxCnt(pflow, pflow->alstate) == txid + 1)) {
```

Comment:
this is missing the obvious case where AppLayerParserGetTxCnt==1, like for TLS. In this case we can trust the TX is relevant.

Reply:
> this is missing the obvious case where AppLayerParserGetTxCnt==1,

I do not think it is missing : in this case txid==0, so  AppLayerParserGetTxCnt==1==txid + 1

And I do not think it is obvious that we can trust the TX is relevant

Reply:
Adding test case in https://github.com/OISF/suricata-verify/pull/2157

Reply:
> > this is missing the obvious case where AppLayerParserGetTxCnt==1,
> 
> I do not think it is missing : in this case txid==0, so AppLayerParserGetTxCnt==1==txid + 1
> 
> And I do not think it is obvious that we can trust the TX is relevant

But this still relies on `de_ctx->force_applayer`, while the idea is that you can just use that one tx

Reply:
> while the idea is that you can just use that one tx

Maybe for TLS, it is ok but in general no you cannot use that one tx by just checking `AppLayerParserGetTxCnt==1` cf https://github.com/OISF/suricata/pull/12166#discussion_r1861811844

signature `alert ip any any -> any any (sid: 1; flow.age: >1; flow: to_server; )`
packets :
- 3 way handshake
- first HTTP2 request
- acked
- after one second, second HTTP2 request begins with a packet with the start of big HTTP2 frame : 
  - parser returns ` AppLayerResult::incomplete`, does not create tx
  - signature triggers
  - the only transaction present is the first one (AppLayerParserGetTxCnt==1) but the first request had all its data before flow.age was 1 second old, so should not be logged

Reply:
And for TLS, I get the argument that you cannot log a wrong transaction as there will forever only be one, but it is not obvious to me that everyone will want the app-layer logged for a signature such as `alert ip any any -> any any (sid: 1; flow.age: >1; flow: to_server; )` if the packet triggering it happens after the clear text TLS handshake...